docs/reprocess-report.md

```diff
@@ -0,0 +1,122 @@
+
+## Case: adding new feature to existing table
+
+Let's use [commit adding `body_simhash` extraction](https://github.com/ooni/pipeline/commit/8e14b20ec368572c0bb831fb958bcc70eb9108a6) as an example. Things to do are the following:
+
+- Add new feature as a nullable column. Adding a `NOT NULL` column will trigger an early table rewrite that is waste of CPU and Disk IO bandwidth.
+- Bump `CODE_VER`
+- Bump `min_compat_code_ver` to the new value of `CODE_VER` for affected "feeders" (`HttpRequestFeeder` and `HttpControlFeeder` in this case)
+- Append new feature columns to `columns`
+- Write code to extract the needed feature for `TheFeeder.row()`, drop those fields in `TheFeeder.pop()` (if needed), test it and deploy.
+- Reprocess all (or the affected) buckets under GNU Make control.
+- Alter the feature column to be `NOT NULL` if needed.
+
+## Marking autoclaved files for reprocessing
+
+The _autoclaved_ files are selected for reingestion and reprocessing based on their `code_ver`. If `autoclaved.code_ver` matches `centrifugation.py:CODE_VER` then the file is skipped altogether (file is not read and decompressed, json is not parsed). If `autoclaved.code_ver` is _compatible_ with `Feeder.min_compat_code_ver` (greater-equal-than) then the corresponding PostgreSQL table is not re-written during a centrifugation pass. So it can be used to reduce amount of burned CPU and database disk IO.
+
+_autoclaved_ file may be marked with `code_ver` equal to 0 (`CODE_VER_REPROCESS`) to force reprocessing of all the feature-tables for this file. There should be no reasons for that besides, maybe, clean-up after a manual database modifications.
+
+Reingestion is different from reprocessing as it may handle changes to autoclaved files themselves and update `autoclaved`, `report`, `measurement` tables accordingly. The easiest way to force reingestion manually is to set `autoclaved.file_sha1` to all-zeros of something like `digest('', 'sha1')`. One of the possible reasons for that is [report deletion](./delete-report.md).
+
+## GNU Make crutch for Airflow
+
+Airflow has an issue in a scheduler, it starts consuming unreasonable amount of resources if there are thousands of _running_ DAGs. So, reprocessing of ≈2300 daily buckets of OONI data has to be micro-managed. One of the usual Linux tools to execute parallel processes is GNU Make, so, it was taken for the [`pipeline-reprocess`](https://github.com/ooni/sysadmin/blob/4defab8e92a2e53e2679a17214162ed058089e7f/scripts/pipeline-reprocess) script.
+
+The way to use the script is the following:
+
+- download it to your $HOME at `datacollector.infra.ooni.io` running Airflow
+- edit `PRJ` with a slug representing a reprocessing session
+- choose a way to list buckets-to-reprocess with `TYPEOF_DEPS`
+- edit `$(PRJ)/...-deps` to reflect the desired logic to select buckets to reprocess
```

Comment:
Is this a necessary step in most cases? Is it wise to have this logic be editing on the fly on the machine and not checked into version control?

Reply:
Yes, it's necessary to keep every `*.tiflag` file in a separate directory. All in all, GNU Make operates on files.

Indeed, there is some value in checking in the edited file into version control to preserve the history, but there is no other value. So, I assume that _committing for historical purposes_ is an assumed action. Otherwise we should write "and commit to version control, and push, and eject the device safely before unplugging, and use `shutdown -h` instead of pulling the cord". All those actions are good and, probably, needed, but those actions are not highly relevant to the discussed matter.

I say "download it to your $HOME at `datacollector.infra.ooni.io` running Airflow" to name the machine and say that it does not need root (and calls `sudo`).